apps/sms/views/conversation/js/conversation.js

```diff
@@ -1178,22 +1168,36 @@ var ConversationView = {
     return draftDiscardPromise.then(() => {
       // Now we'll try to find existing thread for the new message, otherwise
       // let's fallback to new message composer.
-      var threadPromise = Promise.reject();
       if (parameters && parameters.number) {
-        threadPromise = MessageManager.findThreadFromNumber(parameters.number);
+        // A rejected promise will be returned in case we can't find thread
+        // for the specified number.
+        return MessageManager.findThreadFromNumber(parameters.number);
       }
+      return Promise.reject();
+    }).then(
+      (id) => Navigation.toPanel('thread', { id: id, focusComposer: true }),
+      () => {
+        var draftPromise;
+        if (parameters.messageId) {
+          draftPromise = this.storeDraftFromMessage(parameters.messageId);
+        } else if (parameters.number) {
+          var draft = new Draft({
+            recipients: [parameters.number],
+            type: Utils.isEmailAddress(parameters.number) ? 'mms' : 'sms'
+          });
+          draftPromise = Drafts.add(draft).store().then(() => draft.id);
+        } else {
+          draftPromise = Promise.reject(new TypeError('Unknown parameter'));
+        }
 
-      // The rejected promise will be returned in case we can't find thread
-      // for the specified number.
-      return threadPromise.then((id) => {
-        return Navigation.toPanel('thread', { id: id, focusComposer: true });
-      }, () => {
-        return Navigation.toPanel('composer', {
-          activity: parameters,
-          focusComposer: !!(parameters && parameters.number)
-        });
-      });
-    });
+        return draftPromise.then(
+          (draftId) => Navigation.toPanel('composer', {
+            draftId,
+            focusComposer: !!(parameters && parameters.number)
+          })
+        );
+      }
+    );
```

Comment:
I think the original `navigateToComposer` name is misleading since it actually handled both conversation view and composer(new message) view. Maybe we can rename this method to `navigateToNextView` or smth and wrap the reject part from `findThreadFromNumber` with another  `navigateToNewMessage` or `navigateToComposer` for increasing the readability in quick way.